upstream/upkubeless/upstream.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package upkubeless
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/kubeless/kubeless/pkg/client/clientset/versioned"
+	kubelessutil "github.com/kubeless/kubeless/pkg/utils"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/upstream"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+type Upstream struct {
+	kubelessClient versioned.Interface
+}
+
+var _ upstream.Upstream = (*Upstream)(nil)
+
+func MakeUpstream() (*Upstream, error) {
+	kubelessClient, err := kubelessutil.GetKubelessClientOutCluster()
+	if err != nil {
+		return nil, err
+	}
+	return &Upstream{
+		kubelessClient: kubelessClient,
+	}, nil
+}
+
+func (u *Upstream) Roundtrip(app *apps.App, creq *apps.CallRequest, async bool) (io.ReadCloser, error) {
+	name := match(creq.Path, &app.Manifest)
+	if name == "" {
+		return nil, utils.ErrNotFound
+	}
+
+	// Build the JSON request
+	creqData, err := upstream.ServerlessRequestFromCall(creq)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert call into invocation payload")
+	}
+
+	crespData, err := u.invoke(app, name, creq.Path, http.MethodPost, creqData, async)
+	if err != nil {
+		return nil, err
+	}
+
+	return io.NopCloser(bytes.NewReader(crespData)), nil
+}
+
+func FunctionURL(appID apps.AppID, funcName string) (string, error) {
+	clientset := kubelessutil.GetClientOutOfCluster()
+	return functionURL(clientset, appID, funcName)
+}
+
+func functionURL(clientset kubernetes.Interface, appID apps.AppID, funcName string) (string, error) {
+	// Get the function's service URL
+	svc, err := clientset.CoreV1().Services(namespace(appID)).Get(funcName, metav1.GetOptions{})
```

Comment:
Could you please elaborate on the impact of using a namespace here? What is the reasoning behind a per app namespace?

Reply:
Agreed. Reduced all functions (already uniquely named) to `mattermost-kubeless-apps` namespace.

Reply:
Sound good to me:+1: